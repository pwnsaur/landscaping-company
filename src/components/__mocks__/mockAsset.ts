import { Asset } from 'contentful';

export const createMockAsset = (): Asset => {
  return {
    sys: {
      id: '1',
      type: 'Asset',
      createdAt: '2023-04-27T12:00:00Z',
      updatedAt: '2023-04-27T12:00:00Z',
      locale: 'en-US',
      contentType: {
        sys: {
          id: 'asset',
          type: 'Link',
          linkType: 'ContentType',
        },
      },
    },
    fields: {
      title: 'jonathan-kemper-kYl3Yp2hwtQ-unsplash',
      description: '',
      file: {
        url: '//images.ctfassets.net/sandie-clarke-q13Zq1Jufks-unsplash.jpg',
        details: {
          size: 500939,
          image: {
            width: 1920,
            height: 1080,
          },
        },
        fileName: '',
        contentType: 'image/jpeg',
      },
    },
    metadata: {
      tags: [],
    },
    toPlainObject: () => ({} as Asset),
  };
};
