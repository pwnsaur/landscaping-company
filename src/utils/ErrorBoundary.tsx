import React, { ErrorInfo } from 'react';

interface Props {
  children: React.ReactNode;
}

interface State {
  hasError: boolean;
}

class ErrorBoundary extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = { hasError: false };
  }

  static getDerivedStateFromError(_: Error): State {
    return { hasError: true };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    console.log({ error, errorInfo });
  }

  render() {
    if (this.state.hasError) {
      return (
        <div>
          <h1>Lapu neizdevās ielādēt.</h1>
        </div>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
