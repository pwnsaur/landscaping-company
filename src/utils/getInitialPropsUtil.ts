import { AppContext, AppInitialProps } from 'next/app';

import { isMobileUserAgent } from './userAgent';

export async function getInitialPropsUtil(
  appContext: AppContext
): Promise<AppInitialProps & { isMobile: boolean }> {
  let pageProps = {};

  if (appContext.Component.getInitialProps) {
    pageProps = await appContext.Component.getInitialProps(appContext.ctx);
  }

  const userAgent = appContext.ctx.req?.headers['user-agent'] || 'rekt';
  const isMobile = isMobileUserAgent(userAgent);

  return { pageProps, isMobile };
}
