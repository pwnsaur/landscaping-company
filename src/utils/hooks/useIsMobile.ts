import useWindowSize from './useWindowSize';

const useIsMobile = (): boolean => {
  const { width, height } = useWindowSize();
  const currentWidth = width || 0;
  const currentHeight = height || 0;
  const isMobile = currentWidth <= 640;

  return isMobile;
};

export default useIsMobile;
