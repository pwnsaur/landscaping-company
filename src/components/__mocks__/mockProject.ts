import { createMockAsset } from '@components/__mocks__/mockAsset';
import { BLOCKS } from '@contentful/rich-text-types';
import { Entry } from 'contentful';

import { TypeProjectFields } from '@/types/contentfulTypes';

export const createMockProject = (): Entry<TypeProjectFields> => {
  return {
    sys: {
      id: '1',
      type: 'Entry',
      createdAt: '2023-04-27T12:00:00Z',
      updatedAt: '2023-04-27T12:00:00Z',
      locale: 'en-US',
      contentType: {
        sys: {
          id: 'project',
          type: 'Link',
          linkType: 'ContentType',
        },
      },
    },
    fields: {
      title: 'Fifteenth Project',
      slug: 'fifteenth-project',
      content: {
        nodeType: BLOCKS.DOCUMENT,
        data: {},
        content: [],
      },
      coverImage: createMockAsset(),
      date: '2023-04-27',
    },
    metadata: {
      tags: [],
    },
    toPlainObject: () => ({} as Entry<TypeProjectFields>),
    update: () => Promise.resolve({} as Entry<TypeProjectFields>),
  };
};
