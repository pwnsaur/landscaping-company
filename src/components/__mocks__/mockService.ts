import { createMockAsset } from '@components/__mocks__/mockAsset';
import { BLOCKS } from '@contentful/rich-text-types';
import { Entry } from 'contentful';

import { TypeServiceFields } from '@/types/contentfulTypes';

export const createMockService = (): Entry<TypeServiceFields> => {
  return {
    sys: {
      id: '1',
      type: 'Entry',
      createdAt: '2023-04-27T12:00:00Z',
      updatedAt: '2023-04-27T12:00:00Z',
      locale: 'en-US',
      contentType: {
        sys: {
          id: 'service',
          type: 'Link',
          linkType: 'ContentType',
        },
      },
    },
    fields: {
      title: 'Service',
      slug: 'service',
      excerpt: 'service',
      description: {
        nodeType: BLOCKS.PARAGRAPH,
        data: {},
        content: [],
      },
      coverImage: createMockAsset(),
    },
    metadata: {
      tags: [],
    },
    toPlainObject: () => ({} as Entry<TypeServiceFields>),
    update: () => Promise.resolve({} as Entry<TypeServiceFields>),
  };
};
