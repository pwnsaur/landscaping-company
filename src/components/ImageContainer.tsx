import { Asset } from 'contentful';
import Image from 'next/image';
import React, { useEffect, useState, useCallback } from 'react';
import styled from 'styled-components';

import useIsMobile from '@/utils/hooks/useIsMobile';
import ZoomedImage from '@components/ZoomedImage';

type Props = {
  images: Asset[] | undefined;
};

const ImageContainer = ({ images }: Props) => {
  const [zoomedImageIndex, setZoomedImageIndex] = useState<number | null>(null);
  const isMobile = useIsMobile();

  const handleClick = (index: number) => {
    if (!isMobile) {
      setZoomedImageIndex(index);
      document.body.style.overflow = 'hidden';
    }
  };

  const handleClose = useCallback(() => {
    setZoomedImageIndex(null);
    document.body.style.overflow = 'auto';
  }, [setZoomedImageIndex]);

  const navigateImage = useCallback(
    (direction: number) => {
      if (!images || zoomedImageIndex === null) {
        return;
      }

      const newIndex = zoomedImageIndex + direction;
      if (newIndex >= 0 && newIndex < images.length) {
        setZoomedImageIndex(newIndex);
      }
    },
    [images, zoomedImageIndex]
  );

  useEffect(() => {
    const handleKeydown = (e: KeyboardEvent) => {
      if (e.key === 'ArrowLeft') {
        navigateImage(-1);
      } else if (e.key === 'ArrowRight') {
        navigateImage(1);
      } else if (e.key === 'Escape') {
        handleClose();
      }
    };

    if (zoomedImageIndex !== null) {
      window.addEventListener('keydown', handleKeydown);
    }

    return () => {
      window.removeEventListener('keydown', handleKeydown);
    };
  }, [zoomedImageIndex, navigateImage, handleClose]);

  return (
    <>
      <Container>
        {images?.map((image, index) => (
          <ImageWrapper key={image.sys.id} onClick={() => handleClick(index)}>
            <StyledImage
              src={`https:${image.fields.file.url}`}
              alt='project image'
              height={image.fields.file.details.image!.height / 4}
              width={image.fields.file.details.image!.width / 4}
              quality={50}
            />
          </ImageWrapper>
        ))}
      </Container>
      {zoomedImageIndex !== null && !isMobile && images && (
        <ZoomedImage
          src={`https:${images[zoomedImageIndex].fields.file.url}`}
          alt='zoomed image'
          width={images[zoomedImageIndex].fields.file.details.image!.width}
          height={images[zoomedImageIndex].fields.file.details.image!.height}
          close={handleClose}
          previous={() => navigateImage(-1)}
          next={() => navigateImage(1)}
          zoomedImageIndex={zoomedImageIndex}
          imagesLength={images.length}
        />
      )}
    </>
  );
};

export default ImageContainer;

const Container = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(16rem, 1fr));
  width: 100%;
  gap: 1rem;
  justify-content: center;
  margin-top: 2rem;
`;

const StyledImage = styled(Image)`
  position: absolute;
  object-fit: cover;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
`;

const ImageWrapper = styled.div`
  position: relative;
  width: 100%;
  padding-top: 66%;
  overflow: hidden;
  ${({ theme }) => !theme.isMobile && `cursor: pointer;`}
`;
