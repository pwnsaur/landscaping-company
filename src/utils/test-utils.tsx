import { ReactElement, ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';

import { theme } from '@/styles/theme';
import useIsMobile from '@/utils/hooks/useIsMobile';
import { render, RenderOptions } from '@testing-library/react';

interface TestWrapperProps {
  children: ReactNode;
}

const TestWrapper = ({ children }: TestWrapperProps) => {
  const isMobile = useIsMobile();

  return (
    <ThemeProvider theme={{ ...theme, isMobile }}>{children}</ThemeProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>
) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

// eslint-disable-next-line import/export
export * from '@testing-library/react';

// eslint-disable-next-line import/export
export { customRender as render };
