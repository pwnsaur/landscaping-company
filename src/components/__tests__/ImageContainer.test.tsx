import { fireEvent, render, screen } from '@/utils/test-utils';
import ImageContainer from '@/components/ImageContainer';
import useIsMobile from '@/utils/hooks/useIsMobile';
import { createMockAsset } from '../__mocks__/mockAsset';

jest.mock('../../utils/hooks/useIsMobile', () => jest.fn());

const mockImage = createMockAsset();
const mockImages = [
  mockImage,
  { ...mockImage, sys: { ...mockImage.sys, id: '2' } },
];

describe('ImageContainer component', () => {
  test('matches the snapshot', () => {
    const { asFragment } = render(<ImageContainer images={mockImages} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders images', () => {
    (useIsMobile as jest.Mock).mockReturnValue(false);
    render(<ImageContainer images={mockImages} />);
    expect(screen.getAllByAltText('project image')).toHaveLength(2);
  });

  test('clicking on an image when isMobile===false zooms it', () => {
    (useIsMobile as jest.Mock).mockReturnValue(false);
    render(<ImageContainer images={[mockImage]} />);
    fireEvent.click(screen.getByAltText('project image'));
    expect(screen.getByAltText('zoomed image')).toBeInTheDocument();
  });

  test("clicking on an image when isMobile===true doesn't zoom it", () => {
    (useIsMobile as jest.Mock).mockReturnValue(true);
    render(<ImageContainer images={[mockImage]} />);
    fireEvent.click(screen.getByAltText('project image'));
    expect(screen.queryByAltText('zoomed image')).not.toBeInTheDocument();
  });

  test('clcking on the image closes it', () => {
    (useIsMobile as jest.Mock).mockReturnValue(false);
    render(<ImageContainer images={[mockImage]} />);
    fireEvent.click(screen.getByAltText('project image'));
    fireEvent.click(screen.getByAltText('zoomed image'));
    expect(screen.queryByAltText('zoomed image')).not.toBeInTheDocument();
  });
});
