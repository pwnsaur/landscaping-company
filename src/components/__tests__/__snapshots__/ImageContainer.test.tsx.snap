// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageContainer component matches the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: grid;
  grid-template-columns: repeat(auto-fit,minmax(16rem,1fr));
  width: 100%;
  gap: 1rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 2rem;
}

.c2 {
  position: absolute;
  object-fit: cover;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c1 {
  position: relative;
  width: 100%;
  padding-top: 66%;
  overflow: hidden;
  cursor: pointer;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <img
        alt="project image"
        class="c2"
        data-nimg="1"
        decoding="async"
        height="270"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fimages.ctfassets.net%2Fsandie-clarke-q13Zq1Jufks-unsplash.jpg&w=1080&q=50"
        srcset="/_next/image?url=https%3A%2F%2Fimages.ctfassets.net%2Fsandie-clarke-q13Zq1Jufks-unsplash.jpg&w=640&q=50 1x, /_next/image?url=https%3A%2F%2Fimages.ctfassets.net%2Fsandie-clarke-q13Zq1Jufks-unsplash.jpg&w=1080&q=50 2x"
        style="color: transparent;"
        width="480"
      />
    </div>
    <div
      class="c1"
    >
      <img
        alt="project image"
        class="c2"
        data-nimg="1"
        decoding="async"
        height="270"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fimages.ctfassets.net%2Fsandie-clarke-q13Zq1Jufks-unsplash.jpg&w=1080&q=50"
        srcset="/_next/image?url=https%3A%2F%2Fimages.ctfassets.net%2Fsandie-clarke-q13Zq1Jufks-unsplash.jpg&w=640&q=50 1x, /_next/image?url=https%3A%2F%2Fimages.ctfassets.net%2Fsandie-clarke-q13Zq1Jufks-unsplash.jpg&w=1080&q=50 2x"
        style="color: transparent;"
        width="480"
      />
    </div>
  </div>
</DocumentFragment>
`;
